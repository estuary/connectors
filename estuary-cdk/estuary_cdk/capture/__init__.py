from dataclasses import dataclass
from pydantic import Field
from typing import Generic, Awaitable, Any, BinaryIO, Callable
from logging import Logger
import abc
import asyncio
import shutil
import sys
import tempfile
import traceback
import xxhash

from . import request, response
from .. import BaseConnector, Stopped
from ..flow import (
    ConnectorSpec,
    ConnectorState,
    ConnectorStateUpdate,
    EndpointConfig,
    ResourceConfig,
)
from ..pydantic_polyfill import GenericModel


class Request(GenericModel, Generic[EndpointConfig, ResourceConfig, ConnectorState]):
    spec: request.Spec | None = None
    discover: request.Discover[EndpointConfig] | None = None
    validate_: request.Validate[EndpointConfig, ResourceConfig] | None = Field(
        default=None, alias="validate"
    )
    apply: request.Apply[EndpointConfig, ResourceConfig] | None = None
    open: request.Open[EndpointConfig, ResourceConfig, ConnectorState] | None = None
    acknowledge: request.Acknowledge | None = None


class Response(GenericModel, Generic[EndpointConfig, ResourceConfig, ConnectorState]):
    spec: ConnectorSpec | None = None
    discovered: response.Discovered[ResourceConfig] | None = None
    validated: response.Validated | None = None
    applied: response.Applied | None = None
    opened: response.Opened | None = None
    captured: response.Captured | None = None
    checkpoint: response.Checkpoint[ConnectorState] | None = None


@dataclass
class Task:
    """
    Task bundles a capture coroutine and its associated context.

    It facilitates the task in queuing any number of captured documents and
    emitting them upon a checkpoint. Each instance of Task manages an independent
    internal buffer (backed by memory or disk) which is only written to the
    connector's output upon a call to checkpoint(). This allows concurrent
    Tasks to capture consistent checkpoints without trampling one another.

    Task also facilitates logging and graceful stop of a capture coroutine.
    """

    log: Logger
    """Attached Logger of this Task instance, to use for scoped logging."""

    @dataclass
    class Stopping:
        """
        Stopping coordinates the graceful exit of capture Tasks.

        Its `event` is set when Tasks should gracefully stop.
        The Task's coroutine should monitor this event and exit when it's set AND
        it has no more immediate work to do (for example, no further documents are
        currently ready to be captured).
        """

        event: asyncio.Event
        first_error: Exception | None = None
        first_error_task: str | None = None

    stopping: Stopping

    _buffer: tempfile.SpooledTemporaryFile
    _hasher: xxhash.xxh3_128
    _name: str
    _output: BinaryIO
    _tg: asyncio.TaskGroup

    MAX_BUFFER_MEM: int = 1_000_000
    """Maximum amount of memory to use for captured documents between checkpoints,
    before spilling to disk."""

    def __init__(
        self,
        log: Logger,
        name: str,
        output: BinaryIO,
        stopping: Stopping,
        tg: asyncio.TaskGroup,
    ):
        self._buffer = tempfile.SpooledTemporaryFile(max_size=self.MAX_BUFFER_MEM)
        self._hasher = xxhash.xxh3_128()
        self._name = name
        self._output = output
        self._tg = tg
        self.log = log
        self.stopping = stopping

    def captured(self, binding: int, document: Any):
        """Enqueue the document to be captured under the given binding.
        Documents are not actually captured until checkpoint() is called.
        Or, reset() will discard any queued documents."""

        b = Response(
            captured=response.Captured(binding=binding, doc=document)
        ).model_dump_json(by_alias=True, exclude_unset=True)

        self._buffer.write(b.encode())
        self._buffer.write(b"\n")
        self._hasher.update(b)

    def pending_digest(self) -> str:
        """pending_digest returns the digest of captured() documents
        since the last checkpoint() or reset()"""

        return self._hasher.digest().hex()

    def checkpoint(self, state: ConnectorState, merge_patch: bool = True):
        """Emit previously-queued, captured documents follows by a checkpoint"""

        self._emit(
            Response[Any, Any, ConnectorState](
                checkpoint=response.Checkpoint(
                    state=ConnectorStateUpdate(updated=state, mergePatch=merge_patch)
                )
            )
        )

    def reset(self):
        """Discard any captured documents, resetting to an empty state."""
        self._buffer.truncate(0)
        self._buffer.seek(0)
        self._hasher.reset()

    def spawn_child(
        self, name_suffix: str, child: Callable[["Task"], Awaitable[None]]
    ) -> asyncio.Task:
        """
        Spawn a child Task of this Task, using the given name suffix and coroutine.
        The child coroutine will be invoked with a child Task and be polled concurrently.
        """

        child_name = f"{self._name}.{name_suffix}"
        child_log = self.log.getChild(name_suffix)

        async def run_task(parent: Task):
            async with asyncio.TaskGroup() as child_tg:
                try:
                    t = Task(
                        child_log,
                        child_name,
                        parent._output,
                        parent.stopping,
                        child_tg,
                    )
                    await child(t)
                except Exception as exc:
                    child_log.error("".join(traceback.format_exception(exc)))

                    if parent.stopping.first_error is None:
                        parent.stopping.first_error = exc
                        parent.stopping.first_error_task = child_name

                    parent.stopping.event.set()

        task = self._tg.create_task(run_task(self))
        task.set_name(child_name)
        return task

    def _emit(self, response: Response[EndpointConfig, ResourceConfig, ConnectorState]):
        self._buffer.write(
            response.model_dump_json(by_alias=True, exclude_unset=True).encode()
        )
        self._buffer.write(b"\n")
        self._buffer.seek(0)
        shutil.copyfileobj(self._buffer, self._output)
        self._output.flush()
        self.reset()


class BaseCaptureConnector(
    BaseConnector[Request[EndpointConfig, ResourceConfig, ConnectorState]],
    Generic[EndpointConfig, ResourceConfig, ConnectorState],
):
    output: BinaryIO = sys.stdout.buffer

    @abc.abstractmethod
    async def spec(self, log: Logger, _: request.Spec) -> ConnectorSpec:
        raise NotImplementedError()

    @abc.abstractmethod
    async def discover(
        self,
        log: Logger,
        discover: request.Discover[EndpointConfig],
    ) -> response.Discovered[ResourceConfig]:
        raise NotImplementedError()

    @abc.abstractmethod
    async def validate(
        self,
        log: Logger,
        validate: request.Validate[EndpointConfig, ResourceConfig],
    ) -> response.Validated:
        raise NotImplementedError()

    async def apply(
        self,
        log: Logger,
        apply: request.Apply[EndpointConfig, ResourceConfig],
    ) -> response.Applied:
        return response.Applied(actionDescription="")

    @abc.abstractmethod
    async def open(
        self,
        log: Logger,
        open: request.Open[EndpointConfig, ResourceConfig, ConnectorState],
    ) -> tuple[response.Opened, Callable[[Task], Awaitable[None]]]:
        raise NotImplementedError()

    async def acknowledge(self, acknowledge: request.Acknowledge) -> None:
        return None  # No-op.

    async def handle(
        self,
        log: Logger,
        request: Request[EndpointConfig, ResourceConfig, ConnectorState],
    ) -> None:

        if spec := request.spec:
            response = await self.spec(log, spec)
            response.protocol = 3032023
            self._emit(Response(spec=response))

        elif discover := request.discover:
            self._emit(Response(discovered=await self.discover(log, discover)))

        elif validate := request.validate_:
            self._emit(Response(validated=await self.validate(log, validate)))

        elif apply := request.apply:
            self._emit(Response(applied=await self.apply(log, apply)))

        elif open := request.open:
            opened, capture = await self.open(log, open)
            self._emit(Response(opened=opened))

            stopping = Task.Stopping(asyncio.Event())

            async def stop_on_elapsed_interval(interval: int) -> None:
                await asyncio.sleep(interval)
                stopping.event.set()

            # Gracefully exit after the capture interval has elapsed.
            # We don't do this within the TaskGroup because we don't
            # want to block on it.
            asyncio.create_task(stop_on_elapsed_interval(open.capture.intervalSeconds))

            async with asyncio.TaskGroup() as tg:

                task = Task(
                    log.getChild("capture"),
                    "capture",
                    self.output,
                    stopping,
                    tg,
                )
                await capture(task)

            # When capture() completes, the connector exits.
            if stopping.first_error:
                raise Stopped(
                    f"Task {stopping.first_error_task}: {stopping.first_error}"
                )
            else:
                raise Stopped(None)

        elif acknowledge := request.acknowledge:
            await self.acknowledge(acknowledge)

        else:
            raise RuntimeError("malformed request", request)

    def _emit(self, response: Response[EndpointConfig, ResourceConfig, ConnectorState]):
        self.output.write(
            response.model_dump_json(by_alias=True, exclude_unset=True).encode()
        )
        self.output.write(b"\n")
        self.output.flush()
