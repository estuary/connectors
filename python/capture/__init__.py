from queue import Queue
from threading import Thread
import abc
import jsonlines
import orjson
import sys
import typing as t

from .. import BaseConnector, flow
from . import request
from . import response


class Request(t.TypedDict, total=False):
    spec: request.Spec
    discover: request.Discover
    validate: request.Validate
    apply: request.Apply
    open: request.Open
    acknowledge: request.Acknowledge


class Response(t.TypedDict, total=False):
    spec: flow.Spec
    discovered: response.Discovered
    validated: response.Validated
    applied: response.Applied
    opened: response.Opened
    captured: response.Captured
    checkpoint: response.Checkpoint


class Connector(BaseConnector, metaclass=abc.ABCMeta):
    def __init__(self):
        super().__init__()

    @abc.abstractmethod
    def spec(self, spec: request.Spec) -> flow.Spec:
        raise NotImplementedError()

    @abc.abstractmethod
    def discover(self, discover: request.Discover) -> response.Discovered:
        raise NotImplementedError()

    @abc.abstractmethod
    def validate(self, validate: request.Validate) -> response.Validated:
        raise NotImplementedError()

    def apply(self, apply: request.Apply) -> response.Applied:
        return response.Applied()  # No-op.

    @abc.abstractmethod
    def open(self, open: request.Open, emit: t.Callable[[Response], None]) -> None:
        raise NotImplementedError()

    def acknowledge(self, acknowledge: request.Acknowledge) -> None:
        pass  # No-op.

    def _run(self) -> None:
        # Start a read loop in a thread which dispatches to `requests`.
        requests: Queue[Request | None] = Queue(maxsize=8)
        Thread(
            target=_read_loop, name="read loop", args=(requests,), daemon=True
        ).start()

        while True:
            if not (request := requests.get()):
                return  # EOF.
            elif (spec := request.get("spec")) is not None:
                spec_response = self.spec(spec)
                spec_response["protocol"] = 3032023
                self._emit(Response(spec=spec_response))
            elif (discover := request.get("discover")) is not None:
                self._emit(Response(discovered=self.discover(discover)))
            elif (validate := request.get("validate")) is not None:
                self._emit(Response(validated=self.validate(validate)))
            elif (apply := request.get("apply")) is not None:
                self._emit(Response(applied=self.apply(apply)))
            elif (open := request.get("open")) is not None:
                break
            else:
                raise RuntimeError("malformed request", request)

        def _emit_response(response: Response) -> None:
            self._emit(response)

            # Queue rarely has items, so optimize for the common "empty" case.
            # Otherwise non-blocking get() throws Empty, and exceptions are expensive.
            # This is racy and can false-negative but can never false-positive,
            # since we're the only queue consumer.
            if requests.qsize() == 0:
                return

            if not (next := requests.get()):
                return  # EOF
            elif (acknowledge := next.get("acknowledge")) is not None:
                self.acknowledge(acknowledge)
            else:
                raise RuntimeError("malformed request", request)

        self.open(open, _emit_response)


def _read_loop(queue: Queue[Request | None]):
    request: Request
    for request in jsonlines.Reader(sys.stdin, loads=orjson.loads):
        queue.put(request)

    queue.put(None)  # Signal EOF.
