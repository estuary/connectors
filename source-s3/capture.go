package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"

	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/estuary/connectors/go-types/airbyte"
	"github.com/estuary/connectors/go-types/parser"
	"github.com/estuary/connectors/go-types/shardrange"
	log "github.com/sirupsen/logrus"
)

var StreamCompleted = errors.New("stream completed")

type readResult struct {
	err         error
	bucket      string
	streamID    string
	relativeKey string
	state       *objectState
	record      json.RawMessage
}

func NewStream(config *Config, client *s3.S3, configuredStream airbyte.ConfiguredStream, state streamState, shardRange shardrange.Range) (*stream, error) {
	var bucket, prefix = parseStreamName(configuredStream.Stream.Name)
	var matchKey *regexp.Regexp
	var err error
	if config.MatchKeys != "" {
		matchKey, err = regexp.Compile(config.MatchKeys)
		if err != nil {
			return nil, fmt.Errorf("invalid regex for matchKey: %w", err)
		}
	}

	var projections = make(map[string]parser.JsonPointer)
	for k, v := range configuredStream.Projections {
		projections[k] = parser.JsonPointer(v)
	}

	return &stream{
		streamID:         configuredStream.Stream.Name,
		bucket:           bucket,
		prefix:           prefix,
		state:            state,
		client:           client,
		config:           config,
		matchKeys:        matchKey,
		collectionSchema: configuredStream.Stream.JSONSchema,
		projections:      projections,
		shardRange:       shardRange,
	}, nil
}

func (s *stream) Start(ctx context.Context, onResult func(readResult) error) {
	var err = s.captureStreamInternal(ctx, onResult)
	// Send a StreamCompleted to indicate that we've completed successfully
	if err == nil {
		err = StreamCompleted
	}
	onResult(readResult{err: err})
}

type stream struct {
	streamID         string
	bucket           string
	prefix           string
	state            streamState
	client           *s3.S3
	config           *Config
	matchKeys        *regexp.Regexp
	collectionSchema json.RawMessage
	projections      map[string]parser.JsonPointer
	shardRange       shardrange.Range
}

func (s *stream) captureStreamInternal(ctx context.Context, onResult func(readResult) error) error {
	log.Debug("Starting cature")
	var listResults = listAllObjects(ctx, s.bucket, s.prefix, s.client)
	var objectCount = 0
	var importedCount = 0
	for result := range listResults {
		objectCount++
		if result.err != nil {
			return fmt.Errorf("listing objects: %w", result.err)
		}
		if yes, skipRecords := s.shouldImport(result.object); yes {
			if err := s.importObject(ctx, result.object, skipRecords, onResult); err != nil {
				return fmt.Errorf("failed to import object: '%s': %w", result.object.relativeKey, err)
			}
			importedCount++
		}
	}
	log.WithFields(log.Fields{
		"observedObjects": objectCount,
		"importedObjects": importedCount,
	}).Info("completed processing stream")
	return nil
}

func (c *stream) shouldImport(obj *s3Object) (bool, uint64) {
	// Should we skip importing this item because it doesn't match the configured regex?
	if c.matchKeys != nil && !c.matchKeys.MatchString(obj.relativeKey) {
		return false, 0
	}
	// Does this key hash into our shard's assigned range?
	// We hash the bucket and the full key just to be on the safe side.
	var hashKey = obj.bucket + "/" + obj.fullKey()
	if !c.shardRange.IncludesHwHash([]byte(hashKey)) {
		return false, 0
	}

	var prevState = c.state[obj.relativeKey]
	// Does our state indiate that we've already imported this object?
	if prevState != nil && prevState.ETag == obj.etag {
		if prevState.Complete {
			return false, 0
		} else {
			return true, prevState.RecordCount
		}
	}
	return true, 0
}

func (s *stream) importObject(ctx context.Context, obj *s3Object, skipRecords uint64, onResult func(readResult) error) error {
	var fullKey = obj.fullKey()

	var getInput = s3.GetObjectInput{
		Bucket: &obj.bucket,
		Key:    &fullKey,
	}
	var resp, err = s.client.GetObject(&getInput)
	if err != nil {
		return fmt.Errorf("getObject failed: %w", err)
	}
	defer resp.Body.Close()

	var parseConfig = s.makeParseConfig(obj.relativeKey, resp.ContentType, resp.ContentEncoding)
	configFile, err := ioutil.TempFile("", "parser-config-*.json")
	if err != nil {
		return fmt.Errorf("creating parser config temp file: %w", err)
	}
	defer os.Remove(configFile.Name())
	parseConfig.WriteToFile(configFile)

	var docCount uint64 = 0
	err = parser.ParseStream(ctx, configFile.Name(), resp.Body, func(data json.RawMessage) error {
		docCount++
		// Do we need to skip this record because it's already been ingested, as indicated by the
		// offset in the state?
		if docCount > skipRecords {
			return onResult(readResult{
				bucket:      s.bucket,
				streamID:    s.streamID,
				relativeKey: obj.relativeKey,
				record:      data,
				state: &objectState{
					ETag:         *resp.ETag,
					LastModified: *resp.LastModified,
					RecordCount:  docCount,
					Complete:     false,
				},
			})
		}
		return nil
	})
	if err != nil {
		return err
	}

	return onResult(readResult{
		bucket:      s.bucket,
		streamID:    s.streamID,
		relativeKey: obj.relativeKey,
		state: &objectState{
			ETag:         *resp.ETag,
			LastModified: *resp.LastModified,
			RecordCount:  docCount,
			Complete:     true,
		},
	})
}

func (s *stream) makeParseConfig(relativeKey string, contentType *string, contentEncoding *string) *parser.Config {
	var parseConfig = parser.Config{}
	if s.config.Parser != nil {
		parseConfig = s.config.Parser.Copy()
	}

	parseConfig.Filename = relativeKey
	if contentType != nil {
		parseConfig.ContentType = *contentType
	}
	if contentEncoding != nil {
		parseConfig.ContentEncoding = *contentEncoding
	}
	// If the user supplied a location for this, then we'll use that. Otherwise, use the default
	if parseConfig.AddRecordOffset == "" {
		parseConfig.AddRecordOffset = defaultRecordOffsetLocation
	}
	if parseConfig.AddValues == nil {
		parseConfig.AddValues = make(map[parser.JsonPointer]interface{})
	}
	if parseConfig.Schema == nil {
		parseConfig.Schema = s.collectionSchema
	}
	parseConfig.Projections = s.projections
	parseConfig.AddValues[sourceFilenameLocation] = relativeKey
	return &parseConfig
}

const sourceFilenameLocation = "/_meta/source/file"
const defaultRecordOffsetLocation = "/_meta/source/record"

func shouldImport(obj *s3Object, matchKey *regexp.Regexp, state streamState) bool {
	if matchKey != nil && !matchKey.MatchString(obj.relativeKey) {
		return false
	}

	// Have we started importing this file before
	if s, ok := state[obj.relativeKey]; ok {
		if s.ETag == obj.etag {
			return !s.Complete
		}
	}
	return true
}
