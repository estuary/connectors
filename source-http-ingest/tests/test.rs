use anyhow::Context;
use proto_flow::capture::{
    response::{Captured, Opened},
    Response,
};
use reqwest::header::HeaderMap;

use std::process::Stdio;
use tokio::io::AsyncBufReadExt;
use tokio::{
    io::BufReader,
    process::{ChildStdin, ChildStdout, Command},
};

#[tokio::test]
async fn test_discover() {
    let mut child = Command::new(env!("CARGO_BIN_EXE_source-http-ingest"))
        .kill_on_drop(true)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        // This is just easier than having it listen on a random port and then
        // needing to figure out what it is by parsing log output.
        .env("SOURCE_HTTP_INGEST_PORT", "27173")
        .env("LOG_LEVEL", "info")
        .spawn()
        .expect("spawning child process");

    let mut stdin = child.stdin.take().unwrap();
    let mut stdout = BufReader::new(child.stdout.take().unwrap());

    let discover_req = serde_json::json!({
        "discover": {
            "config": {}
        }
    });
    write_capture_request(&discover_req, &mut stdin)
        .await
        .expect("failed to wrtie req");

    let mut buf = String::new();
    stdout
        .read_line(&mut buf)
        .await
        .expect("failed to read stdout");
    let result: serde_json::Value = serde_json::from_str(&buf).expect("deserialize response");
    insta::assert_json_snapshot!(result);
}

#[tokio::test]
async fn test_http_request_processing() -> Result<(), anyhow::Error> {
    let mut child = Command::new(env!("CARGO_BIN_EXE_source-http-ingest"))
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        // This is just easier than having it listen on a random port and then
        // needing to figure out what it is by parsing log output.
        .env("SOURCE_HTTP_INGEST_PORT", "27172")
        .env("LOG_LEVEL", "info")
        .spawn()?;

    let stdin = child.stdin.take().unwrap();
    let stdout = child.stdout.take().unwrap();
    let result = run_http_request_processing(stdin, stdout).await;
    let kill_result = child.kill().await;
    result?;
    kill_result?;
    Ok(())
}

async fn run_http_request_processing(
    mut stdin: ChildStdin,
    stdout: ChildStdout,
) -> anyhow::Result<()> {
    let mut reader = BufReader::new(stdout);

    // Write open and expect to read opened.
    write_capture_request(&open_json(), &mut stdin).await?;
    match read_response(&mut reader).await.context("reading opened")? {
        Response {
            opened: Some(Opened {
                explicit_acknowledgements,
            }),
            ..
        } => assert!(explicit_acknowledgements),
        _ => anyhow::bail!("unexpected repsonse type, expected opened"),
    };
    // The connector sends the opened response before it actually starts listening
    // on the port. This is an (admittedly hacky) way of dealing with that.
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    // Use two separate clients to fire requests off in separate background tasks.
    let client_a = reqwest::ClientBuilder::new().build()?;
    let client_b = client_a.clone();

    let headers_a = HeaderMap::new();
    let mut headers_b = HeaderMap::new();
    headers_b.insert(
        "X-Webhook-Id",
        reqwest::header::HeaderValue::from_str("same-id").unwrap(),
    );

    let handle_a = tokio::task::spawn(write_docs(
        128,
        "http://localhost:27172/aliceCo/test/webhook-data".to_string(),
        client_a,
        headers_a,
    ));
    let handle_b = tokio::task::spawn(write_docs(
        128,
        "http://localhost:27172/another.json".to_string(),
        client_b,
        headers_b,
    ));

    // We'll re-use this value to write acknowledgements as we read commits
    let ack_request = serde_json::json!({"acknowledge": {}});

    // Expect to read 256 documents in total, each followed immediately by a checkpoint.
    // As checkpoints are read, write an ack to unblock responses.
    let mut a_docs = 0;
    let mut b_docs = 0;
    for i in 0..512 {
        let expect_commit = i % 2 == 1;
        let expect_document = !expect_commit;

        let response = read_response(&mut reader).await?;
        let Response {
            checkpoint,
            captured,
            ..
        } = &response;

        match (checkpoint.as_ref(), captured.as_ref()) {
            (Some(_), None) if expect_commit => {
                write_capture_request(&ack_request, &mut stdin)
                    .await
                    .context("writing acknowledge")?;
            }
            (None, Some(Captured { binding, doc_json })) if expect_document => {
                if *binding == 0 {
                    a_docs += 1;
                } else if *binding == 1 {
                    b_docs += 1;
                } else {
                    anyhow::bail!(
                        "got document with unexpected binding {binding}: {}",
                        doc_json
                    );
                }
            }
            _ => {
                anyhow::bail!(
                    "got unexpected response on round {i}, expect_commit: {expect_commit}: {:?}",
                    response
                );
            }
        }
    }

    // ensure that the background tasks all completed successfully
    handle_a.await.context("result_a")??;
    handle_b.await.context("result_b")??;

    anyhow::ensure!(a_docs == 128, "expected 128 a_docs, got: {a_docs}");
    anyhow::ensure!(b_docs == 128, "expected 128 b_docs, got: {b_docs}");

    Ok(())
}

async fn read_response(stdout: &mut BufReader<ChildStdout>) -> anyhow::Result<Response> {
    let mut line = String::with_capacity(256);

    tokio::time::timeout(
        std::time::Duration::from_secs(10),
        stdout.read_line(&mut line),
    )
    .await??;
    if line.is_empty() {
        anyhow::bail!("stdout line was empty");
    }
    let resp = serde_json::from_str::<Response>(&line)?;
    Ok(resp)
}

async fn write_docs(
    count: usize,
    url: String,
    client: reqwest::Client,
    headers: HeaderMap,
) -> anyhow::Result<()> {
    let expected = serde_json::json!({"published": 1});
    for i in 0..count {
        if let Err(err) = post_doc(i, url.as_str(), &client, &expected, headers.clone()).await {
            eprintln!("failed to post doc at index: {i}: error: {:?}", err);
            anyhow::bail!("failed to post document at index: {i}: error: {:?}", err);
        }
    }
    Ok(())
}

async fn post_doc(
    i: usize,
    url: &str,
    client: &reqwest::Client,
    expected_resp: &serde_json::Value,
    headers: HeaderMap,
) -> anyhow::Result<()> {
    let resp = client
        .post(url)
        .headers(headers)
        .json(&serde_json::json!({"docIndex": i, "url": url}))
        .timeout(std::time::Duration::from_secs(10))
        .send()
        .await?
        .error_for_status()?
        .json::<serde_json::Value>()
        .await?;
    if &resp != expected_resp {
        anyhow::bail!("actual response: '{resp}' differed from expected: '{expected_resp}'");
    }
    Ok(())
}

async fn write_capture_request(
    req: &serde_json::Value,
    stdin: &mut ChildStdin,
) -> anyhow::Result<()> {
    use tokio::io::AsyncWriteExt;

    let buf = serde_json::to_vec(req).unwrap();

    stdin.write_all(&buf).await?;
    stdin.write_all(b"\n").await?;
    Ok(())
}

fn open_json() -> serde_json::Value {
    serde_json::json!({
        "open": {
            "version": "canary-version",
            "range": {
                "keyBegin": 0,
                "keyEnd": 4294967295i64,
            },
            "state_json": "{}",
            "capture": {
                "name": "aliceCo/test/ingest",
                "config": {},
                "intervalSeconds": 10,
                "bindings": [
                    {
                        "collection": {
                            "name": "aliceCo/test/webhook-data",
                            "write_schema_json": {
                                "type": "object",
                                "properties": {
                                    "_meta": {
                                        "type": "object",
                                        "properties": { "webhookId": {"type": "string"}},
                                        "required": ["webhookId"]
                                    }
                                },
                                "required": ["_meta"]
                            },
                            "key": ["/_meta/webhookId"],
                            "partitionFields": [],
                            "projections": []
                        },
                        "resourceConfig": {},
                        "resourcePath": ["aliceCo/test/webhook-data"]
                    },
                    {
                        "collection": {
                            "name": "aliceCo/test/webhook-data",
                            "write_schema_json": {
                                "type": "object",
                                "properties": {
                                    "_meta": {
                                        "type": "object",
                                        "properties": { "webhookId": {"type": "string"}},
                                        "required": ["webhookId"]
                                    }
                                },
                                "required": ["_meta"]
                            },
                            "key": ["/_meta/webhookId"],
                            "partitionFields": [],
                            "projections": []
                        },
                        "resourceConfig": {
                            "path": "another.json",
                            "idFromHeader": "X-Webhook-Id"
                        },
                        "resourcePath": ["/another.json"]
                    }
                ]
            }
        }
    })
}
