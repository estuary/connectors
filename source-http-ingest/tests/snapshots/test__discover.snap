---
source: tests/test.rs
expression: result
---
{
  "discovered": {
    "bindings": [
      {
        "documentSchema": {
          "properties": {
            "_meta": {
              "description": "These fields are automatically added by the connector, and do not need to be specified in the request body",
              "properties": {
                "headers": {
                  "additionalProperties": {
                    "type": "string"
                  },
                  "description": "HTTP headers that were sent with the request will get added here. Headers that are known to be sensitive or not useful will not be included",
                  "type": "object"
                },
                "receivedAt": {
                  "description": "Timestamp of when the request was received by the connector",
                  "format": "date-time",
                  "type": "string"
                },
                "webhookId": {
                  "description": "The id of the webhook request, which is automatically added by the connector",
                  "type": "string"
                }
              },
              "required": [
                "webhookId",
                "receivedAt"
              ],
              "type": "object"
            }
          },
          "required": [
            "_meta"
          ],
          "type": "object",
          "x-infer-schema": true
        },
        "key": [
          "/_meta/webhookId"
        ],
        "recommendedName": "webhook-data",
        "resourceConfig": {
          "idFromHeader": null,
          "path": null,
          "stream": null
        },
        "resourcePath": [
          "webhook-data"
        ]
      }
    ]
  }
}
